packages/components/menu/src/menu.ts

```diff
@@ -412,20 +425,92 @@ export default defineComponent({
 
       const ulStyle = useMenuCssVar(props, 0)
 
-      const vMenu = h(
-        'ul',
-        {
-          key: String(props.collapse),
-          role: 'menubar',
-          ref: menu,
-          style: ulStyle.value,
-          class: {
-            [nsMenu.b()]: true,
-            [nsMenu.m(props.mode)]: true,
-            [nsMenu.m('collapse')]: props.collapse,
+      const recusiveMouseInSubMenu = (
+        slot: VNodeArrayChildren | VNodeChildAtom
+      ): boolean => {
+        if (
+          !slot ||
+          (isVNode(slot) && (slot.type as Component).name !== 'ElSubMenu') ||
+          typeof slot !== 'object'
+        )
+          return false
+
+        if (Array.isArray(slot)) {
+          for (const element of slot) {
+            const value = recusiveMouseInSubMenu(element)
+            if (value) return true
+          }
+
+          return false
+        }
+
+        if (slot.component?.exposed?.mouseInChild.value) return true
+
+        const subMenuSlots = slot.component?.slots?.default?.()
+
+        if (!subMenuSlots || !subMenuSlots.entries()) return false
+
+        for (const [i, subMenuSlot] of subMenuSlots.entries()) {
+          if ((subMenuSlot.type as Component).name === 'ElSubMenu') {
+            const value = recusiveMouseInSubMenu(subMenuSlots[i])
+            if (value) return value
+          }
+        }
+
+        return false
+      }
+
+      const constructIndexPath = (index: string) => {
+        const indexPath: string[] = []
+        const splits = index.split('-')
+
+        for (let i = 0; i < splits.length; i++) {
+          const nextIndexPath = splits.slice(0, i + 1).join('-')
+          indexPath.push(nextIndexPath)
+        }
+
+        return indexPath
+      }
+
+      const directives: DirectiveArguments = props.collapseOnClickOutside
+        ? [
+            [
+              vClickoutside,
+              () => {
+                const hasMouseInMenu = recusiveMouseInSubMenu(slot)
+
+                if (!hasMouseInMenu) {
+                  timeout?.()
+                  ;({ stop: timeout } = useTimeoutFn(() => {
+                    openedMenus.value.forEach((openedMenu) => {
+                      emit('close', openedMenu, constructIndexPath(openedMenu))
+                    })
```

Comment:
I don't think there is a need to emit each `close` event.

Reply:
If `menu-trigger` is set to `hover` and you open 2 sub-menus, it will trigger to emit `close` twice. To be consistent with that flow, I decided to emit a `close` event for all opened sub-menus.

Reply:
https://github.com/element-plus/element-plus/assets/102006695/308c3d80-b956-4092-a6f4-237e99798ac9